### PR TITLE
1.21.09 Update

### PR DESCRIPTION
**Pull request by CameronGoodwin**
1.21.09 Update contains the following improvements:

- Switch to DateTimeOffset from community feedback
- Added support for RefreshAfter value for UserStoreId to refresh the ClawbackQueue UserPurchaseIds every 2 weeks
- Added support for UnmanagedConsumables
- Added sandbox support
- Added support for Pass product type (Store Managed Subscriptions)
- Updated response ASCII text to be more readable
- Updates to Controllers to support and work with the Client sample

### CODE REVIEW

File: MicrosoftStoreServicesSample/ClawbackManager/ClawbackManager.cs [hunk 1]
```diff
@@ -116,6 +116,41 @@ private async Task RemoveClawbackQueueItemFromTrackingIdAsync(string trackingId,
...
+            {
+                using (var dbContext = ServerDBController.CreateDbContext(_config, cV, _logger))
+                {
+                    //  We should have a unique LineItemId
+                    var item = dbContext.ClawbackQueue.Find(TrackingId);
+                    
+                    if (item != null)
+                    {
+                        item.UserPurchaseId = NewKeyPurchaseId;
+                        await dbContext.SaveChangesAsync();
```
Comment (scm-xbox) [inline MicrosoftStoreServicesSample/ClawbackManager/ClawbackManager.cs:138]: I don't think this SaveChangesAsync is required since it's called regardless below.

File: MicrosoftStoreServicesSample/Controllers/CollectionsController.cs [hunk 2]
```diff
@@ -109,6 +109,18 @@ public async Task<ActionResult<string>> Query([FromBody] ClientCollectionsQueryR
                 return response.ToString();
             }
 
+            bool includeJson = false;
+            if (Request.Headers.ContainsKey("User-Agent"))
+            {
+                string userAgent = Request.Headers["User-Agent"];
+                if (!string.IsNullOrEmpty(userAgent) && userAgent.Equals("Microsoft.StoreServicesClientSample"))
```
Comment (scm-xbox) [inline MicrosoftStoreServicesSample/Controllers/CollectionsController.cs:116]: Prefer == to .Equals unless you need a case-insensitive compare.

File: MicrosoftStoreServicesSample/Controllers/CollectionsController.cs [hunk 3]
```diff
@@ -123,6 +135,21 @@ public async Task<ActionResult<string>> Query([FromBody] ClientCollectionsQueryR
             };
             queryRequest.Beneficiaries.Add(beneficiary);
 
+            if (!string.IsNullOrEmpty(clientRequest.sbx))
+            {
+                queryRequest.SandboxId = clientRequest.sbx;
+            }
+
+            if (clientRequest.EntitlementFilters != null &&
```
Comment (scm-xbox) [inline MicrosoftStoreServicesSample/Controllers/CollectionsController.cs:143]: Prefer null coalescing. 

```csharp
clientRequest.EntitlementFilters?.Count > 0
```

File: MicrosoftStoreServicesSample/Controllers/CollectionsController.cs [hunk 4]
```diff
@@ -144,34 +171,69 @@ public async Task<ActionResult<string>> Query([FromBody] ClientCollectionsQueryR
...
-                                          item.ProductKind,
-                                          item.ProductId,
-                                          satisfyingEntitlements.ToString(),
-                                          AcquisitionTypeFriendlyName(item.AcquisitionType),
-                                          consumableInfo.ToString());
+                //  If this is from the Client sample, include the JSON so that it can display the items in the UI
+                //  properly
+                if (includeJson)
+                {
+                    response.Append("RawResponse:");
```
Comment (scm-xbox) [inline MicrosoftStoreServicesSample/Controllers/CollectionsController.cs:233]: Should there be a space after the colon? or a newline between them or anything? Formatting of this might look weird.

File: MicrosoftStoreServicesSample/Controllers/PurchaseController.cs [hunk 5]
```diff
@@ -450,7 +513,6 @@ private static string FormatResponseFromClawbackActionList(List<ClawbackActionIt
         {
             var response = new StringBuilder("");
             response.Append(
-                    "|--------------------------------------------------------------------------------------------------------------------------------------------------------------------|\n" +
                     "| Action State | ProductId    | Quantity | LineItemState | Purchased Date         | Refunded Date          | Candidate Id     | Candidate Consume Date | LineItemId                           |\n" +
```
Comment (scm-xbox) [inline MicrosoftStoreServicesSample/Controllers/PurchaseController.cs:516]: Not part of this change, but you may want to consider switching to AppendFormat using width operators `{0,20}` and such...

https://docs.microsoft.com/en-us/dotnet/standard/base-types/composite-formatting#alignment-component

File: MicrosoftStoreServicesSample/ClawbackManager/ClawbackManager.cs (hunk 1 above)
Comment (CameronGoodwin) [reply MicrosoftStoreServicesSample/ClawbackManager/ClawbackManager.cs:138]: removed the save towards the bottom so it only happens if it needs to.

File: MicrosoftStoreServicesSample/Controllers/CollectionsController.cs (hunk 2 above)
Comment (CameronGoodwin) [reply MicrosoftStoreServicesSample/Controllers/CollectionsController.cs:116]: OK

File: MicrosoftStoreServicesSample/Controllers/CollectionsController.cs (hunk 3 above)
Comment (CameronGoodwin) [reply MicrosoftStoreServicesSample/Controllers/CollectionsController.cs:143]: I would prefer to keep it as it is for readability on this specific one.

File: MicrosoftStoreServicesSample/Controllers/CollectionsController.cs (hunk 4 above)
Comment (CameronGoodwin) [reply MicrosoftStoreServicesSample/Controllers/CollectionsController.cs:233]: sure, it won't affect the current client parsing the data.
